hbase-server/src/main/java/org/apache/hadoop/hbase/master/procedure/SnapshotVerifyProcedure.java

```diff
@@ -0,0 +1,206 @@
+  or more contributor license agreements.  See the NOTICE file
+  distributed with this work for additional information
+  regarding copyright ownership.  The ASF licenses this file
+  to you under the Apache License, Version 2.0 (the
+  "License"); you may not use this file except in compliance
+  with the License.  You may obtain a copy of the License at
+  http://www.apache.org/licenses/LICENSE-2.0
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+ */
+
+package org.apache.hadoop.hbase.master.procedure;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import org.apache.hadoop.hbase.DoNotRetryIOException;
+import org.apache.hadoop.hbase.ServerName;
+import org.apache.hadoop.hbase.client.RegionInfo;
+import org.apache.hadoop.hbase.procedure2.Procedure;
+import org.apache.hadoop.hbase.procedure2.ProcedureStateSerializer;
+import org.apache.hadoop.hbase.procedure2.ProcedureSuspendedException;
+import org.apache.hadoop.hbase.procedure2.ProcedureYieldException;
+import org.apache.hadoop.hbase.procedure2.RemoteProcedureDispatcher;
+import org.apache.hadoop.hbase.regionserver.SnapshotVerifyCallable;
+import org.apache.hadoop.hbase.snapshot.CorruptedSnapshotException;
+import org.apache.yetus.audience.InterfaceAudience;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.apache.hadoop.hbase.shaded.protobuf.ProtobufUtil;
+import org.apache.hadoop.hbase.shaded.protobuf.generated.MasterProcedureProtos.SnapshotVerifyParameter;
+import org.apache.hadoop.hbase.shaded.protobuf.generated.MasterProcedureProtos.SnapshotVerifyProcedureStateData;
+import org.apache.hadoop.hbase.shaded.protobuf.generated.SnapshotProtos.SnapshotDescription;
+
+/**
+ *  A remote procedure which is used to send verify snapshot request to region server.
+ */
+@InterfaceAudience.Private
+public class SnapshotVerifyProcedure
+    extends ServerRemoteProcedure implements ServerProcedureInterface {
+  private static final Logger LOG = LoggerFactory.getLogger(SnapshotProcedure.class);
+  private SnapshotDescription snapshot;
+  private List<RegionInfo> regions;
+  private int expectedNumRegion;
+
+  public SnapshotVerifyProcedure() {}
+
+  public SnapshotVerifyProcedure(SnapshotDescription snapshot, List<RegionInfo> regions,
+      ServerName targetServer, int expectedNumRegion) {
+    this.targetServer = targetServer;
+    this.snapshot = snapshot;
+    this.regions = regions;
+    this.expectedNumRegion = expectedNumRegion;
+  }
+
+  @Override
+  protected void complete(MasterProcedureEnv env, Throwable error) {
+    if (error != null) {
+      Throwable realError = error.getCause();
```

Comment:
Why we need this trick here? We will always wrap the actual exception?

Reply:
yes, on the master side we will wrap this into a RemoteProcedureException.